diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/chapter3/CollapseIfStatementsRule.kt

```diff
@@ -0,0 +1,161 @@
+import org.cqfn.diktat.common.config.rules.RulesConfig
+import org.cqfn.diktat.common.config.rules.getRuleConfig
+import org.cqfn.diktat.ruleset.constants.Warnings
+import org.cqfn.diktat.ruleset.rules.DiktatRule
+import org.cqfn.diktat.ruleset.utils.KotlinParser
+
+import com.pinterest.ktlint.core.ast.ElementType.BINARY_EXPRESSION
+import com.pinterest.ktlint.core.ast.ElementType.BLOCK_COMMENT
+import com.pinterest.ktlint.core.ast.ElementType.EOL_COMMENT
+import com.pinterest.ktlint.core.ast.ElementType.IF
+import com.pinterest.ktlint.core.ast.ElementType.KDOC
+import com.pinterest.ktlint.core.ast.ElementType.LBRACE
+import com.pinterest.ktlint.core.ast.ElementType.OPERATION_REFERENCE
+import com.pinterest.ktlint.core.ast.ElementType.RBRACE
+import com.pinterest.ktlint.core.ast.ElementType.THEN
+import com.pinterest.ktlint.core.ast.ElementType.WHITE_SPACE
+import com.pinterest.ktlint.core.ast.children
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.psi.KtBlockExpression
+import org.jetbrains.kotlin.psi.KtIfExpression
+
+import java.util.Stack
+
+typealias placeOfWarningForCurrentNode = Pair<Int, ASTNode>
+
+/**
+ * Rule for redundant nested if-statements, which could be collapsed into a single one
+ */
+class CollapseIfStatementsRule(configRules: List<RulesConfig>) : DiktatRule(
+    "collapse-if",
+    configRules,
+    listOf(
+        Warnings.COLLAPSE_IF_STATEMENTS
+    )
+) {
+    private val configuration by lazy {
+        CollapseIfStatementsConfiguration(
+            configRules.getRuleConfig(Warnings.COLLAPSE_IF_STATEMENTS)?.configuration ?: emptyMap()
+        )
+    }
+
+    // We hold the warnings, which we raised, since in case of multi nested if-statement,
+    // there are could be several identical warning for one line
+    private val listOfWarnings: MutableSet<placeOfWarningForCurrentNode> = mutableSetOf()
+
+    override fun logic(node: ASTNode) {
+        if (node.elementType == IF) {
+            process(node)
+        }
+    }
+
+    private fun process(node: ASTNode) {
+        val startCollapseFromLevel = configuration.startCollapseFromNestedLevel
+        val listOfNestedNodes: Stack<ASTNode> = Stack()
+
+        var nestedIfNode = findNestedIf(node)
+        while (nestedIfNode != null) {
+            listOfNestedNodes.push(nestedIfNode)
+            nestedIfNode = findNestedIf(nestedIfNode)
+        }
+        val nestedLevel = listOfNestedNodes.size + 1
+        if (nestedLevel < startCollapseFromLevel) {
+            return
+        }
+        while (listOfNestedNodes.isNotEmpty()) {
+            val currNode = listOfNestedNodes.pop()
+            // Since the external `if` statement is not the direct parent,
+            // we need multiple steps to take the required one
+            // BLOCK -> THEN -> IF
+            val currParentNode = currNode.treeParent.treeParent.treeParent
+            if (listOfWarnings.add(currNode.startOffset to currNode)) {
```

Comment:
Is keeping node in the list necessary too? If we are in lint mode, then offsets won't change and will be unique for node, and in fix mode both offsets and nodes are changing